text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,128 @@
+- CNB Issue: (leave blank)
+- Supersedes: "N/A"
+
+# Summary
+[summary]: #summary
+
+A set of predefined URIs for referring to buildpacks from various sources (protocols). 
+
+_Please note that the scheme denotes protocol and not type._
+
+# Motivation
+[motivation]: #motivation
+
+Given that we support a multitude of sources for buildpacks, a certain level of ambiguity has naturally presented
+itself. We currently _try_ to resolve the various sources through a complex set of rules yet the ambiguity to users
+still exists. In config files, this ambiguity is at times resolved by introducing a separate key name to signify
+the source (protocol).
+
+The aim of this RFC is to propose a set of URI schemes that could be used in both CLI arguments and config files
+while solving the ambiguity problem.
+
+A few references to further express the problem:
+
+- [A poll](https://buildpacks.slack.com/archives/C94UJCNV6/p1580920669081100) to understand what users expect when
+providing buildpacks and there is overlap of sources.
+- [Issue](https://github.com/buildpacks/pack/issues/489) to provide additional clarity in logs due to ambiguity. 
+- [A conversation](https://buildpacks.slack.com/archives/CJ6B92ZSB/p1578502454017000) when trying to determine how to
+provide reference to all buildpacks in builder.
+- [Issue](https://github.com/buildpacks/pack/issues/376) where an `image` key was added to differentiate between
+directory and image. 
+ 
+# What it is
+[what-it-is]: #what-it-is
+
+### Scheme
+
+| Short Name | Format | Examples |
+|--- |--- |--- |
+| Filesystem | `file://[<host>]/<path>` | `file:///./my/buildpack.tgz`<br>`file:///home/user/my/buildpack.tgz`
+| URL | `http[s]://<host>/<path>` | `http://example.com/my/buildpack.tgz`<br>`https://example.com/my/buildpack.tgz`  
+| Docker | `docker://[<host>]/<path>[:<tag>]` | `docker:///ubuntu:latest`<br>`docker://gcr.io/distroless/nodejs` 
+| CNB Builder | `cnb:builder://[<id>[@<version>]]` | `cnb:builder://`<br>`cnb:builder://bp.id`<br>`cnb:builder://bp.id@bp.version`
+
+
+### Filesystem
+
+The filesystem scheme is the implementation of [RFC 8089](https://tools.ietf.org/html/rfc8089).
+
+### URL
+
+The URL scheme is the implementation of [RFC 7230](https://tools.ietf.org/html/rfc7230#section-2.7.1).
+
+### Docker
+
+The docker scheme denotes the use of the [Docker HTTP API v2 protocol](https://docs.docker.com/registry/spec/api/).
+
+- `<host>` is optional and defaults to `index.docker.io`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `docker:/ubuntu:latest`
+
+### CNB Builder
+
+- `cnb:builder://` - A reference to ALL the buildpacks in the builder.
+- `cnb:builder://bp.id` - A reference to a buildpack with id `bp.id` in the builder. If there is more than one version
+available, a error should be thrown.
+- `cnb:builder://bp.id@bp.version` - A reference to a buildpack with id `bp.id` at the specific version `bp.version`.
```

Comment:
> A non-empty scheme component followed by a colon (:), consisting of a sequence of characters beginning with a letter and followed by any combination of letters, digits, plus (+), period (.), or hyphen (-). Although schemes are case-insensitive, the canonical form is lowercase and documents that specify schemes must do so with lowercase letters. Examples of popular schemes include http, https, ftp, mailto, file, data, and irc. URI schemes should be registered with the Internet Assigned Numbers Authority (IANA), although non-registered schemes are used in practice.[b]

This suggests that our scheme should be `cnb-builder`.

Reply:
Wikipedia is a loose form of a source. :D

Here's an example of a valid URI that has been spec'd containing multiple colons:

[URN](https://tools.ietf.org/html/rfc8141#section-2):
> A Uniform Resource Name (URN) is a Uniform Resource Identifier (URI)
   that is assigned under the "urn" URI scheme and a particular URN
   namespace, with the intent that the URN will be a persistent,
   location-independent resource identifier.

```
urn:example:1/406/47452/2
```

Reply:
This section might also help: https://tools.ietf.org/html/rfc3986#section-1.1.2

Reply:
And another: https://tools.ietf.org/html/rfc3986#section-3

![image](https://user-images.githubusercontent.com/475559/74876326-13d96400-5329-11ea-997e-3ec4a94cd84e.png)

Reply:
https://tools.ietf.org/html/rfc3986#section-3.1
> scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )

>  Scheme names consist of a sequence of characters beginning with a
   letter and followed by any combination of letters, digits, plus
   ("+"), period ("."), or hyphen ("-").  Although schemes are case-
   insensitive, the canonical form is lowercase and documents that
   specify schemes must do so with lowercase letters.  An implementation
   should accept uppercase letters as equivalent to lowercase in scheme
   names (e.g., allow "HTTP" as well as "http") for the sake of
   robustness but should only produce lowercase scheme names for

Reply:
BTW, I may be wrong here, but I think we're discussing a URI (vs a URN which does not indicate how to access a resources); if that's true the scheme is a bit tighter than the UR_N_ scheme.

Reply:
Maybe I'm being pedantic, but I just find it hard to viusally differentiate between the scheme and the authority if the scheme has colons in it. To my eyes, that first colon indicates a seperation between the scheme and the rest of the URI, but that's just me.